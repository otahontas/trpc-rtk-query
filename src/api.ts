import { type ApiEndpointQuery } from "@reduxjs/toolkit/query";
import {
  type Api,
  type BaseQueryFn,
  type CreateApiOptions,
  type EndpointDefinitions,
  createApi as createRTKQueryApi,
} from "@reduxjs/toolkit/query/react";
import { type AnyRouter } from "@trpc/server";

import { type CreateEndpointDefinitions } from "./create-endpoint-definitions";
import { type TRPCBaseQuery, createTRPCBaseQuery } from "./create-trpc-base-query";
import { type AnyApi, type SupportedModule } from "./rtk-types";
import { type TRPCClientOptions } from "./trpc-client-options";
import { type DisabledEndpointOptions, wrapApiToProxy } from "./wrap-api-to-proxy";

/*
 * Creates a new rtk api that exposes endpoints and react hooks generated from trpc
 * types.
 */
export const createApi = <
  TRouter extends AnyRouter,
  ReducerPath extends string = "api",
  TagTypes extends string = never,
  Definitions extends EndpointDefinitions = CreateEndpointDefinitions<
    TRouter,
    TRPCBaseQuery,
    ReducerPath,
    TagTypes
  >,
>(
  options: TRPCClientOptions<TRouter> &
    Omit<
      CreateApiOptions<TRPCBaseQuery, Definitions, ReducerPath, TagTypes>,
      "baseQuery" | "endpoints" // endpoints and baseQuery are generated and injected by this library "endpoints"
    > & {
      endpointOptions?: {
        [K in keyof Definitions]?: Omit<Definitions[K], DisabledEndpointOptions>;
      };
    },
) =>
  wrapApiToProxy({
    api: createRTKQueryApi({
      ...options,
      baseQuery: createTRPCBaseQuery(options),
      // We're injecting endpoints later with proxy, but need to cast them
      // beforehand for proper typings to be exposed to users
      endpoints: () => ({}) as Definitions,
    }),
    endpointOptions: options.endpointOptions,
    useQueryFunction: false,
  });

/**
 * Generic type for api that has injectEndpoint method for run time injection and
 * endpoints for reading previously generated definitions.
 * @internal
 **/
type InjectableWithEndpoints = Pick<AnyApi, "endpoints" | "injectEndpoints">;

/*
 * Enhances existing api with endpoints and react hooks generated from trpc types.
 */
export const enhanceApi = <
  TRouter extends AnyRouter,
  ExistingApi extends InjectableWithEndpoints,
  // == "Save" the types needed to build up proper new api type to type variables ==
  // 1. Current baseQuery from existing api
  BaseQuery extends
    BaseQueryFn = ExistingApi["endpoints"][keyof ExistingApi["endpoints"]]["Types"]["BaseQuery"],
  // 2. Endpoints record values mapped to their inner definitions
  Endpoints = {
    [Endpoint in keyof ExistingApi["endpoints"]]: ExistingApi["endpoints"][Endpoint] extends ApiEndpointQuery<
      infer EndpointDefinition,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any
    >
      ? EndpointDefinition
      : never;
  },
  // 3. Reducer path
  ReducerPath extends
    string = ExistingApi["endpoints"][keyof ExistingApi["endpoints"]]["Types"]["ReducerPath"],
  // 4. Tag types
  TagTypes extends
    string = ExistingApi["endpoints"][keyof ExistingApi["endpoints"]]["Types"]["TagTypes"],
  // New definitions
  NewDefinitions extends EndpointDefinitions = CreateEndpointDefinitions<
    TRouter,
    BaseQuery,
    ReducerPath,
    TagTypes
  >,
>(
  options: TRPCClientOptions<TRouter> & {
    api: ExistingApi;
  } & {
    endpointOptions?: {
      [K in keyof NewDefinitions]?: Omit<NewDefinitions[K], DisabledEndpointOptions>;
    };
  },
) =>
  wrapApiToProxy({
    api: options.api as Api<
      BaseQuery,
      Endpoints & CreateEndpointDefinitions<TRouter, BaseQuery, ReducerPath, TagTypes>,
      ReducerPath,
      TagTypes,
      SupportedModule
    >,
    endpointOptions: options.endpointOptions,
    tRPCClientOptions: options,
    useQueryFunction: true,
  });
