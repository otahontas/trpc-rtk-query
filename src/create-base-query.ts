import { type BaseQueryFn } from "@reduxjs/toolkit/query/react";
import {
  TRPCClientError,
  type TRPCRequestOptions,
  TRPCUntypedClient,
  getUntypedClient,
} from "@trpc/client";
import { type AnyRouter, TRPCError } from "@trpc/server";
import { getHTTPStatusCodeFromError } from "@trpc/server/http";

import { CreateTRPCApiClientOptions } from "./create-trpc-api-client-options";

/* Errors baseQuery can return. Follows the conventions of RTK query's fetchBaseQuery */
export type TRPCBaseQueryError =
  | {
      /**
       * * `"TRPC_CLIENT_ERROR"`:
       *   An error that happened on trpc client. Original error is stringified in error
       *   attribute.
       **/
      data?: undefined;
      error: string;
      message: string;
      name: string;
      status: "TRPC_CLIENT_ERROR";
    }
  | {
      /**
       * * `"TRPC_ERROR"`:
       *   An error that was returned by trpc backend. Original error is stringified in
       *   error attribute.
       **/
      data?: undefined;
      error: string;
      message: string;
      name: string;
      status: "TRPC_ERROR";
      statusCode: number;
    }
  | {
      /**
       * * `"UNKNOWN_ERROR"`:
       *   A unknown error type that captures error not wrapped by trpc. Original error
       *   is stringified in error attribute
       **/
      data?: unknown;
      error: string;
      status: "UNKNOWN_ERROR";
    };

/** Helper type to differentiate between whether client is ready to be used or should
 *  be grabbed from callback
 *  @internal
 **/
type ClientStatus<TRouter extends AnyRouter> =
  | {
      client: TRPCUntypedClient<TRouter>;
      ready: true;
    }
  | {
      getClient: NonNullable<CreateTRPCApiClientOptions<TRouter>["getClient"]>;
      ready: false;
    };

/** Resolve how to get the trpc client
 * @internal
 **/
const resolveClientStatus = <TRouter extends AnyRouter>(
  createTRPCApiClientOptions: CreateTRPCApiClientOptions<TRouter>,
): ClientStatus<TRouter> => {
  if ("client" in createTRPCApiClientOptions) {
    return {
      client: getUntypedClient<TRouter>(createTRPCApiClientOptions.client),
      ready: true,
    };
  } else if ("getClient" in createTRPCApiClientOptions) {
    return {
      getClient: createTRPCApiClientOptions.getClient,
      ready: false,
    };
  }
  throw new Error("No client or getClient specified in options!");
};

/** Typings for base query that (not previously existing) api created with
 * createTRPCApi uses.
 **/
export type BaseQueryForTRPCClient = BaseQueryFn<
  // Arguments for baseQuery. Should be used when injecting endpoints
  {
    procedureArguments: unknown;
    procedurePath: string;
    procedureType: "mutation" | "query";
  },
  // Result type. Endpoint definitions are typed manually instead of deriving them from
  // this type, so it's enough that this is unknown.
  unknown,
  // Use typed errors
  TRPCBaseQueryError,
  // Allow request options (e.g context and signal) to be passed in via extra options
  TRPCRequestOptions,
  // No meta is returned
  never
>;

/** Create a base query that uses trpc client under the hood.
 **/
export const createBaseQueryForTRPCClient = <TRouter extends AnyRouter>(
  createTRPCApiClientOptions: CreateTRPCApiClientOptions<TRouter>,
): BaseQueryForTRPCClient => {
  const clientStatus = resolveClientStatus(createTRPCApiClientOptions);

  return async (baseQueryArguments, baseQueryApi, extraOptions) => {
    try {
      const { procedureArguments, procedurePath, procedureType } = baseQueryArguments;
      const resolvedClient = clientStatus.ready
        ? clientStatus.client
        : getUntypedClient<TRouter>(await clientStatus.getClient(baseQueryApi));
      return {
        data: await resolvedClient[procedureType](
          procedurePath,
          procedureArguments,
          extraOptions,
        ),
      };
    } catch (error) {
      let properlyShapedError: {
        error: TRPCBaseQueryError;
      };
      if (error instanceof TRPCClientError) {
        properlyShapedError = {
          error: {
            error: String(error),
            message: error.message,
            name: error.name,
            status: "TRPC_CLIENT_ERROR",
          },
        };
      } else if (error instanceof TRPCError) {
        properlyShapedError = {
          error: {
            error: String(error),
            message: error.message,
            name: error.name,
            status: "TRPC_ERROR",
            statusCode: getHTTPStatusCodeFromError(error),
          },
        };
      } else {
        properlyShapedError = {
          error: {
            error: String(error),
            status: "UNKNOWN_ERROR",
          },
        };
      }
      return properlyShapedError;
    }
  };
};
