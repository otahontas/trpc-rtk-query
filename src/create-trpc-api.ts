import { type ApiEndpointQuery } from "@reduxjs/toolkit/dist/query/core/module"; // TODO: don't import from dist
import {
  type Api,
  type BaseQueryFn,
  // type CreateApiOptions,
  createApi,
} from "@reduxjs/toolkit/query/react";
import { type AnyRouter } from "@trpc/server";

import {
  type BaseQueryForTRPCClient,
  createBaseQueryForTRPCClient,
} from "./create-base-query";
import { type CreateEndpointDefinitionsFromTRPCRouter } from "./create-endpoint-definitions";
import { type CreateTRPCApiClientOptions } from "./create-trpc-api-client-options";
import {
  type Injectable,
  type SupportedModule,
  wrapApiToProxy,
} from "./wrap-api-to-proxy";

// type NonAllowedApiOptions = Extract<
//   // eslint-disable-next-line @typescript-eslint/no-explicit-any
//   keyof CreateApiOptions<any, any, any, any>,
//   "baseQuery" | "endpoints"
// >;

/*
 * Creates a new api using trpc under the hood
 */
export const createTRPCApi = <TRouter extends AnyRouter>(
  options: CreateTRPCApiClientOptions<TRouter>,
) => {
  const nonProxyApi = createApi({
    baseQuery: createBaseQueryForTRPCClient(options),
    // We're injecting endpoints later with proxy, but need to cast them
    // beforehand for proper typings to be exposed to users
    endpoints: () =>
      ({}) as CreateEndpointDefinitionsFromTRPCRouter<
        TRouter,
        BaseQueryForTRPCClient,
        "api",
        never
      >,
  });
  return wrapApiToProxy({
    nonProxyApi,
    useQueryFunction: false,
  });
};

type InjectableWithEndpoints = Injectable & {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  endpoints: Record<string, ApiEndpointQuery<any, any>>;
};

// TODO: Allow passing in settings for api (reducerpath, tagtypes etc)
export type InjectTRPCEndpointsToApiOptions<
  TRouter extends AnyRouter,
  ExistingApi extends InjectableWithEndpoints,
> = CreateTRPCApiClientOptions<TRouter> & {
  existingApi: ExistingApi;
};

export const injectTRPCEndpointsToApi = <
  TRouter extends AnyRouter,
  ExistingApi extends InjectableWithEndpoints,
  // == "Save" the types needed to build up proper new api type to type variables ==
  // Current baseQuery from existing api
  BaseQuery extends
    BaseQueryFn = ExistingApi["endpoints"][keyof ExistingApi["endpoints"]]["Types"]["BaseQuery"],
  // Endpoints record values mapped to their inner definitions
  Endpoints = {
    [Endpoint in keyof ExistingApi["endpoints"]]: ExistingApi["endpoints"][Endpoint] extends ApiEndpointQuery<
      infer EndpointDefinition,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any
    >
      ? EndpointDefinition
      : never;
  },
  // Reducer path
  ReducerPath extends
    string = ExistingApi["endpoints"][keyof ExistingApi["endpoints"]]["Types"]["ReducerPath"],
  // Tag types
  TagTypes extends
    string = ExistingApi["endpoints"][keyof ExistingApi["endpoints"]]["Types"]["TagTypes"],
>(
  options: InjectTRPCEndpointsToApiOptions<TRouter, ExistingApi>,
) => {
  const nonProxyApi = options.existingApi as unknown as Api<
    BaseQuery,
    Endpoints &
      CreateEndpointDefinitionsFromTRPCRouter<
        TRouter,
        BaseQuery,
        ReducerPath,
        TagTypes
      >,
    ReducerPath,
    TagTypes,
    SupportedModule
  >;

  return wrapApiToProxy({
    createTrpcApiClientOptions: options,
    nonProxyApi,
    useQueryFunction: true,
  });
};
