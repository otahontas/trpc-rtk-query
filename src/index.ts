export { createApi, enhanceApi } from "./api";
