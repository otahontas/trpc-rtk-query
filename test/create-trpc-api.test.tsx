import { configureStore } from "@reduxjs/toolkit";
import {
  type BaseQueryApi,
  createApi as createRTKQueryApi,
} from "@reduxjs/toolkit/query/react";
import { createTRPCProxyClient } from "@trpc/client";
import { createHTTPServer } from "@trpc/server/adapters/standalone";
import { setTimeout } from "node:timers/promises";
import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { beforeAll, describe, expect, it, vi } from "vitest";

import { createTRPCApi, injectTRPCEndpointsToApi } from "../src/create-trpc-api";
import {
  type AppRouter,
  appRouter,
  testClientOptions,
  testPort,
  userFixtures,
} from "./fixtures";

type TestServer = {
  close: () => Promise<void>;
};

export const startTestServer = (): Promise<TestServer> =>
  new Promise((resolveCreate) => {
    const { server } = createHTTPServer({
      router: appRouter,
    });

    const closePromise = (): Promise<void> =>
      new Promise((resolveClose, rejectClose) => {
        server.close((error) => {
          if (error) {
            console.error("Failed to close test server!");
            rejectClose(error);
            return;
          }
          resolveClose();
        });
      });

    server.listen(testPort, () => {
      console.log("Test server listening!");
      resolveCreate({
        close: closePromise,
      });
    });
  });

// helper for rendering component to json for snapshots
const renderedToJSon = (component: renderer.ReactTestRenderer) => {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON;
};

// Client side options that can be passed in when testing
const existingApiTestQueryArgument = "giveMeData";
const existingApiTestQuerySuccessResponse = {
  returning: "data",
};
const existingApiTestQueryFailureResponse = {
  error: "bad stuff happened",
};

// helper that allows creating new api instances when needed
const createRTKQueryApiLazily = () =>
  createRTKQueryApi({
    baseQuery: (baseQueryArguments: { getResponseArgument: string }) => {
      if (baseQueryArguments.getResponseArgument !== existingApiTestQueryArgument) {
        return existingApiTestQueryFailureResponse;
      }
      return {
        data: existingApiTestQuerySuccessResponse,
      };
    },
    endpoints: (builder) => ({
      getResponse: builder.query<string, string>({
        query: (getResponseArgument: string) => ({
          getResponseArgument,
        }),
      }),
    }),
  });

describe("create-trpc-api", () => {
  describe.each([
    {
      testCase: "using api generated by createTRPCApi",
    },
    {
      apiForCreateApiOptions: {
        createRTKQueryApiLazily,
      },
      testCase: "using pre made api",
    },
  ])("when $testCase and ", ({ apiForCreateApiOptions }) => {
    describe.each([
      {
        getApi: () => {
          const base = {
            client: createTRPCProxyClient<AppRouter>(testClientOptions),
          };
          if (apiForCreateApiOptions?.createRTKQueryApiLazily) {
            const existingApi = createRTKQueryApiLazily();
            return injectTRPCEndpointsToApi<AppRouter, typeof existingApi>({
              ...base,
              existingApi,
            });
          }
          return createTRPCApi(base);
        },
        testCase: "using passed client",
      },
      {
        getApi: () => {
          const base = {
            getClient: async (baseQueryApi: BaseQueryApi) => {
              // Check that correct apiArgs object is passed
              expect(baseQueryApi.type).toBeDefined();
              expect(baseQueryApi.endpoint).toBeDefined();

              // Return proxy client
              return createTRPCProxyClient<AppRouter>(testClientOptions);
            },
          };
          if (apiForCreateApiOptions?.createRTKQueryApiLazily) {
            const existingApi = createRTKQueryApiLazily();
            return injectTRPCEndpointsToApi<AppRouter, typeof existingApi>({
              ...base,
              existingApi,
            });
          }
          return createTRPCApi(base);
        },
        testCase: "using getClient to get the client",
      },
    ])("when $testCase", ({ getApi }) => {
      it("Generates an api instance", () => {
        const api = getApi();
        expect(api).toBeDefined();
      });

      it("Generates queries", () => {
        const api = getApi();
        const { useGetUserByIdQuery, useListUsersQuery } = api;
        expect(useGetUserByIdQuery).toBeDefined();
        expect(useListUsersQuery).toBeDefined();
      });

      it("Generates mutations", () => {
        const api = getApi();
        const { useCreateUserMutation, useUpdateNameMutation } = api;
        expect(useUpdateNameMutation).toBeDefined();
        expect(useCreateUserMutation).toBeDefined();
      });

      it.each([
        "useQuery",
        "useQueryState",
        "useQuerySubscription",
        "useLazyQuery",
        "useLazyQuerySubscription",
      ] as const)(
        "Generates %s hook when accessing hooks through endpoints[endpoint] property",
        (queryName) => {
          const api = getApi();
          const query = api.endpoints.getUserById[queryName];
          expect(query).toBeDefined();
        },
      );

      it("Generates usePrefetch", () => {
        const api = getApi();
        const { usePrefetch } = api;
        expect(usePrefetch).toBeDefined();
      });

      it("Generates hook for deeply nested route", () => {
        const api = getApi();
        const { useNested_Deep_GetVeryNestedMessageQuery } = api;
        expect(useNested_Deep_GetVeryNestedMessageQuery).toBeDefined();
      });

      it("Generates hook for deeply nested routes through endpoints[endpoint]", () => {
        const api = getApi();
        const query =
          api.endpoints.nested_Deep_GetVeryNestedMessage.useQuerySubscription;
        expect(query).toBeDefined();
      });

      describe("and making actual requests with hooks renders correctly", () => {
        // Create react app for testing. Keeping this in scope so it's easier to type.
        const createReactTestApp = (getApiForStore: typeof getApi) => {
          const api = getApiForStore();
          const store = configureStore({
            middleware: (getDefaultMiddleware) => [
              ...getDefaultMiddleware(),
              api.middleware,
            ],
            reducer: {
              [api.reducerPath]: api.reducer,
            },
          });

          const createComponentWrapper = (Component: () => JSX.Element) =>
            renderer.create(
              <Provider store={store}>
                <Component />
              </Provider>,
            );
          return {
            api,
            createComponentWrapper,
            store,
          };
        };

        // how much to wait for loading state to resolve
        // TODO: migrate to waitFor when adding react testing library
        const msToWaitBeforeRenderingWithoutLoadingState = 150;

        beforeAll(async () => {
          const { close } = await startTestServer();
          console.log("waiting a bit before running integration tests");
          await setTimeout(250);
          return () => close();
        });

        it("with successful useUserIdQuery", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const Component = () => {
            const { useGetUserByIdQuery } = api;
            const userId = 1;
            const { data, error, isLoading } = useGetUserByIdQuery(userId);
            if (isLoading) {
              return <div>Loading...</div>;
            }
            if (error || !data) {
              return <div>Error</div>;
            }
            expect(data).toEqual(userFixtures[1]);
            return (
              <div>
                <p>Id: {data.id}</p>
                <p>Name: {data.name}</p>
              </div>
            );
          };
          const app = createComponentWrapper(Component);
          // first render
          let result = renderedToJSon(app);
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          expect(result).toMatchSnapshot();
          result = renderedToJSon(app);
          expect(JSON.stringify(result)).not.toContain("Loading...");
          expect(JSON.stringify(result)).not.toContain("Error");
          expect(result).toMatchSnapshot();
        });

        it("with failing useUserIdQuery", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const Component = () => {
            const { useGetUserByIdQuery } = api;
            const userId = 4;
            // TODO: errors should be properly typed from basequery!
            const { data, error, isLoading } = useGetUserByIdQuery(userId);
            if (isLoading) {
              return <div>Loading...</div>;
            }
            if (error || !data) {
              return <div>Error</div>;
            }
            expect(data).toEqual(userFixtures[1]);
            return (
              <div>
                <p>Id: {data.id}</p>
                <p>Name: {data.name}</p>
              </div>
            );
          };
          const app = createComponentWrapper(Component);
          // first render
          let result = renderedToJSon(app);
          expect(result).toMatchSnapshot();
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          // result after data has loaded and component has re-rendered
          result = renderedToJSon(app);
          expect(JSON.stringify(result)).not.toContain("Loading...");
          expect(JSON.stringify(result)).toContain("Error");
          expect(result).toMatchSnapshot();
        });

        it("with successful deep nested query", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const Component = () => {
            const { useNested_Deep_GetVeryNestedMessageQuery } = api;
            const myInput = "heyoooo";
            const { data, error, isLoading } = useNested_Deep_GetVeryNestedMessageQuery(
              {
                deepInput: myInput,
              },
            );
            if (isLoading) {
              return <div>Loading...</div>;
            }
            if (error || !data) {
              return <div>Error</div>;
            }
            expect(data.inputBack).toStrictEqual(myInput);
            return (
              <div>
                <p>inputBack: {data.inputBack}</p>
                <p>messageFromDeep: {data.messageFromDeep}</p>
              </div>
            );
          };
          const app = createComponentWrapper(Component);
          // first render
          let result = renderedToJSon(app);
          expect(result).toMatchSnapshot();
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          // result after data has loaded and component has re-rendered
          result = renderedToJSon(app);
          expect(JSON.stringify(result)).not.toContain("Loading...");
          expect(JSON.stringify(result)).not.toContain("Error");
          expect(result).toMatchSnapshot();
        });

        it("with call to usePrefetch", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const userId = 1;
          const PrefetcherComponent = () => {
            const { usePrefetch } = api;
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore ignore for now, for runtime testing. Fix type later
            const prefetch = usePrefetch("getUserById");
            expect(prefetch).toBeDefined();
            prefetch(userId);
            return <>prefetched</>;
          };
          const QueryComponent = () => {
            const { useQueryState } = api.endpoints.getUserById;
            const { data } = useQueryState(userId);
            expect(data).toEqual(userFixtures[1]);
            return (
              <div>
                <p>Id: {data!.id}</p>
                <p>Name: {data!.name}</p>
              </div>
            );
          };
          const Parent = () => {
            const [showQueryComponent, setShowQueryComponent] = React.useState(false);
            const handleMouseEnter = () => {
              setShowQueryComponent(true);
            };
            return (
              <div onMouseEnter={handleMouseEnter}>
                {showQueryComponent ? <QueryComponent /> : <PrefetcherComponent />}
              </div>
            );
          };
          // render prefetcher component
          const component = createComponentWrapper(Parent);
          let result = renderedToJSon(component);
          expect(JSON.stringify(result)).toContain("prefetched");
          expect(result).toMatchSnapshot();
          // wait for data to load
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          // manually trigger the callback
          result.props["onMouseEnter"]();
          // render query component and check that prefetch was actually called
          // properly through proxy (e.g. query state should be loaded)
          result = renderedToJSon(component);
          expect(JSON.stringify(result)).not.toContain("Error");
          expect(JSON.stringify(result)).toContain("Id");
          expect(JSON.stringify(result)).toContain("Name");
          expect(result).toMatchSnapshot();
        });

        it("does not inject endpoints again", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          // Try to listen error log from rtk
          const consoleErrorMock = vi
            .spyOn(console, "error")
            .mockImplementation(() => {});
          const FirstComponent = () => {
            const { useListUsersQuery } = api;
            const { data } = useListUsersQuery();
            return <>{JSON.stringify(data)}</>;
          };
          const SecondComponent = () => {
            const { useListUsersQuery } = api;
            const { data } = useListUsersQuery();
            return <>{JSON.stringify(data)}</>;
          };
          const Parent = () => {
            return (
              <>
                <FirstComponent />
                <SecondComponent />
              </>
            );
          };
          const app = createComponentWrapper(Parent);
          // first render
          renderedToJSon(app);
          expect(consoleErrorMock).not.toHaveBeenCalled();
        });
      });

      it("Can be extended with enchangeEndpoints", () => {
        const api = getApi();
        expect(api.enhanceEndpoints).toBeDefined();

        // Running this to make sure we haven't broken anything api side
        api.enhanceEndpoints({
          addTagTypes: ["User"],
          endpoints: {
            getUserById: {
              onQueryStarted() {},
              providesTags: ["User"],
            },
            updateName: {
              onQueryStarted() {},
            },
          },
        });
      });
    });
  });

  it("doesn't replace previous hooks when passing in and existing api", () => {
    const existingApi = createRTKQueryApiLazily();
    const client = createTRPCProxyClient<AppRouter>(testClientOptions);
    const api = injectTRPCEndpointsToApi({
      client,
      existingApi,
    });
    const { useListUsersQuery } = api;
    expect(useListUsersQuery).toBeDefined();
    const { useGetResponseQuery } = api;
    expect(useGetResponseQuery).toBeDefined();
  });
});
