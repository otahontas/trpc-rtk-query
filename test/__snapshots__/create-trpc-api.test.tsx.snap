// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with call to usePrefetch 1`] = `
<div
  onMouseEnter={[Function]}
>
  prefetched
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with call to usePrefetch 2`] = `
<div
  onMouseEnter={[Function]}
>
  <div>
    <p>
      Id: 
      1
    </p>
    <p>
      Name: 
      John Johnny
    </p>
  </div>
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 2`] = `
<div>
  Error
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful deep nested query 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful deep nested query 2`] = `
<div>
  <p>
    inputBack: 
    heyoooo
  </p>
  <p>
    messageFromDeep: 
    Hello from deep
  </p>
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 2`] = `
<div>
  <p>
    Id: 
    1
  </p>
  <p>
    Name: 
    John Johnny
  </p>
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with call to usePrefetch 1`] = `
<div
  onMouseEnter={[Function]}
>
  prefetched
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with call to usePrefetch 2`] = `
<div
  onMouseEnter={[Function]}
>
  <div>
    <p>
      Id: 
      1
    </p>
    <p>
      Name: 
      John Johnny
    </p>
  </div>
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 2`] = `
<div>
  Error
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful deep nested query 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful deep nested query 2`] = `
<div>
  <p>
    inputBack: 
    heyoooo
  </p>
  <p>
    messageFromDeep: 
    Hello from deep
  </p>
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using api generated by createTRPCApi' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 2`] = `
<div>
  <p>
    Id: 
    1
  </p>
  <p>
    Name: 
    John Johnny
  </p>
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with call to usePrefetch 1`] = `
<div
  onMouseEnter={[Function]}
>
  prefetched
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with call to usePrefetch 2`] = `
<div
  onMouseEnter={[Function]}
>
  <div>
    <p>
      Id: 
      1
    </p>
    <p>
      Name: 
      John Johnny
    </p>
  </div>
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 2`] = `
<div>
  Error
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful deep nested query 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful deep nested query 2`] = `
<div>
  <p>
    inputBack: 
    heyoooo
  </p>
  <p>
    messageFromDeep: 
    Hello from deep
  </p>
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using getClient to get the client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 2`] = `
<div>
  <p>
    Id: 
    1
  </p>
  <p>
    Name: 
    John Johnny
  </p>
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with call to usePrefetch 1`] = `
<div
  onMouseEnter={[Function]}
>
  prefetched
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with call to usePrefetch 2`] = `
<div
  onMouseEnter={[Function]}
>
  <div>
    <p>
      Id: 
      1
    </p>
    <p>
      Name: 
      John Johnny
    </p>
  </div>
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with failing useUserIdQuery 2`] = `
<div>
  Error
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful deep nested query 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful deep nested query 2`] = `
<div>
  <p>
    inputBack: 
    heyoooo
  </p>
  <p>
    messageFromDeep: 
    Hello from deep
  </p>
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`create-trpc-api > when 'using pre made api' and  > when 'using passed client' > and making actual requests with hooks renders correctly > with successful useUserIdQuery 2`] = `
<div>
  <p>
    Id: 
    1
  </p>
  <p>
    Name: 
    John Johnny
  </p>
</div>
`;
